File: public/app/features/explore-map/crdt/hlc.ts
/**
 * Hybrid Logical Clock (HLC) implementation
 *
 * Combines logical time (Lamport timestamp) with physical wall-clock time
 * to provide timestamps that are:
 * - Monotonically increasing
 * - Causally consistent
 * - Approximately synchronized with real time
 *
 * Used for conflict resolution in LWW-Register CRDTs.
 */

export interface HLCTimestamp {
  logicalTime: number;  // Lamport timestamp component
  wallTime: number;     // Physical clock component (milliseconds since epoch)
  nodeId: string;       // Unique node identifier for tie-breaking
}

export class HybridLogicalClock {
  private logicalTime: number = 0;
  private lastWallTime: number = 0;
  private nodeId: string;

  constructor(nodeId: string) {
    this.nodeId = nodeId;
    this.lastWallTime = Date.now();
  }

  /**
   * Advance the clock for a local event
   * Returns a new timestamp representing "now"
   */
  tick(): HLCTimestamp {
    const now = Date.now();

    if (now > this.lastWallTime) {
      // Physical clock advanced - use it
      this.lastWallTime = now;
      this.logicalTime = 0;
    } else {
      // Physical clock hasn't advanced - increment logical component
      this.logicalTime++;
    }

    return this.clone();
  }

  /**
   * Update clock based on received timestamp from another node
   * Ensures causal consistency: if A → B, then timestamp(A) < timestamp(B)
   */
  update(receivedTimestamp: HLCTimestamp): void {
    const now = Date.now();
    const maxWall = Math.max(this.lastWallTime, receivedTimestamp.wallTime, now);

    if (maxWall === this.lastWallTime && maxWall === receivedTimestamp.wallTime) {
      // Same wall time - take max logical time and increment
      this.logicalTime = Math.max(this.logicalTime, receivedTimestamp.logicalTime) + 1;
    } else if (maxWall === receivedTimestamp.wallTime) {
      // Received timestamp has newer wall time
      this.lastWallTime = maxWall;
      this.logicalTime = receivedTimestamp.logicalTime + 1;
    } else {
      // Our wall time or physical clock is newer
      this.lastWallTime = maxWall;
      this.logicalTime = 0;
    }
  }

  /**
   * Get current timestamp without advancing the clock
   */
  now(): HLCTimestamp {
    return this.clone();
  }

  /**
   * Create a copy of the current timestamp
   */
  clone(): HLCTimestamp {
    return {
      logicalTime: this.logicalTime,
      wallTime: this.lastWallTime,
      nodeId: this.nodeId,
    };
  }

  /**
   * Get the node ID
   */
  getNodeId(): string {
    return this.nodeId;
  }
}

/**
 * Compare two HLC timestamps
 * Returns:
 *   < 0 if a < b
 *   > 0 if a > b
 *   = 0 if a == b
 */
export function compareHLC(a: HLCTimestamp, b: HLCTimestamp): number {
  // First compare wall time
  if (a.wallTime !== b.wallTime) {
    return a.wallTime - b.wallTime;
  }

  // Then compare logical time
  if (a.logicalTime !== b.logicalTime) {
    return a.logicalTime - b.logicalTime;
  }

  // Finally compare node IDs for deterministic tie-breaking
  return a.nodeId.localeCompare(b.nodeId);
}

/**
 * Check if timestamp a happened before timestamp b
 */
export function happensBefore(a: HLCTimestamp, b: HLCTimestamp): boolean {
  return compareHLC(a, b) < 0;
}

/**
 * Check if timestamp a happened after timestamp b
 */
export function happensAfter(a: HLCTimestamp, b: HLCTimestamp): boolean {
  return compareHLC(a, b) > 0;
}

/**
 * Check if two timestamps are equal
 */
export function timestampEquals(a: HLCTimestamp, b: HLCTimestamp): boolean {
  return compareHLC(a, b) === 0;
}

/**
 * Get the maximum of two timestamps
 */
export function maxTimestamp(a: HLCTimestamp, b: HLCTimestamp): HLCTimestamp {
  return compareHLC(a, b) >= 0 ? a : b;
}

/**
 * Serialize timestamp to JSON
 */
export function serializeHLC(timestamp: HLCTimestamp): string {
  return JSON.stringify(timestamp);
}

/**
 * Deserialize timestamp from JSON
 */
export function deserializeHLC(json: string): HLCTimestamp {
  return JSON.parse(json);
}

File: pkg/services/exploremap/crdt/hlc.go
package crdt

import (
	"encoding/json"
	"time"
)

// HLCTimestamp represents a hybrid logical clock timestamp
type HLCTimestamp struct {
	LogicalTime int64  `json:"logicalTime"`
	WallTime    int64  `json:"wallTime"` // milliseconds since epoch
	NodeID      string `json:"nodeId"`
}

// HybridLogicalClock manages hybrid logical time for a node
type HybridLogicalClock struct {
	logicalTime  int64
	lastWallTime int64
	nodeID       string
}

// NewHybridLogicalClock creates a new HLC with the given node ID
func NewHybridLogicalClock(nodeID string) *HybridLogicalClock {
	return &HybridLogicalClock{
		logicalTime:  0,
		lastWallTime: time.Now().UnixMilli(),
		nodeID:       nodeID,
	}
}

// Tick advances the clock for a local event and returns a new timestamp
func (h *HybridLogicalClock) Tick() HLCTimestamp {
	now := time.Now().UnixMilli()

	if now > h.lastWallTime {
		// Physical clock advanced - use it
		h.lastWallTime = now
		h.logicalTime = 0
	} else {
		// Physical clock hasn't advanced - increment logical component
		h.logicalTime++
	}

	return h.Clone()
}

// Update updates the clock based on a received timestamp from another node
func (h *HybridLogicalClock) Update(received HLCTimestamp) {
	now := time.Now().UnixMilli()
	maxWall := max(h.lastWallTime, received.WallTime, now)

	if maxWall == h.lastWallTime && maxWall == received.WallTime {
		// Same wall time - take max logical time and increment
		h.logicalTime = max(h.logicalTime, received.LogicalTime, 0) + 1
	} else if maxWall == received.WallTime {
		// Received timestamp has newer wall time
		h.lastWallTime = maxWall
		h.logicalTime = received.LogicalTime + 1
	} else {
		// Our wall time or physical clock is newer
		h.lastWallTime = maxWall
		h.logicalTime = 0
	}
}

// Clone returns a copy of the current timestamp
func (h *HybridLogicalClock) Clone() HLCTimestamp {
	return HLCTimestamp{
		LogicalTime: h.logicalTime,
		WallTime:    h.lastWallTime,
		NodeID:      h.nodeID,
	}
}

// Now returns the current timestamp without advancing the clock
func (h *HybridLogicalClock) Now() HLCTimestamp {
	return h.Clone()
}

// GetNodeID returns the node ID
func (h *HybridLogicalClock) GetNodeID() string {
	return h.nodeID
}

// CompareHLC compares two HLC timestamps
// Returns: -1 if a < b, 0 if a == b, 1 if a > b
func CompareHLC(a, b HLCTimestamp) int {
	// First compare wall time
	if a.WallTime != b.WallTime {
		if a.WallTime < b.WallTime {
			return -1
		}
		return 1
	}

	// Then compare logical time
	if a.LogicalTime != b.LogicalTime {
		if a.LogicalTime < b.LogicalTime {
			return -1
		}
		return 1
	}

	// Finally compare node IDs for deterministic tie-breaking
	if a.NodeID < b.NodeID {
		return -1
	} else if a.NodeID > b.NodeID {
		return 1
	}

	return 0
}

// HappensBefore checks if timestamp a happened before timestamp b
func HappensBefore(a, b HLCTimestamp) bool {
	return CompareHLC(a, b) < 0
}

// HappensAfter checks if timestamp a happened after timestamp b
func HappensAfter(a, b HLCTimestamp) bool {
	return CompareHLC(a, b) > 0
}

// TimestampEquals checks if two timestamps are equal
func TimestampEquals(a, b HLCTimestamp) bool {
	return CompareHLC(a, b) == 0
}

// MaxTimestamp returns the later of two timestamps
func MaxTimestamp(a, b HLCTimestamp) HLCTimestamp {
	if CompareHLC(a, b) >= 0 {
		return a
	}
	return b
}

// MarshalJSON implements json.Marshaler
func (t HLCTimestamp) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		LogicalTime int64  `json:"logicalTime"`
		WallTime    int64  `json:"wallTime"`
		NodeID      string `json:"nodeId"`
	}{
		LogicalTime: t.LogicalTime,
		WallTime:    t.WallTime,
		NodeID:      t.NodeID,
	})
}

// UnmarshalJSON implements json.Unmarshaler
func (t *HLCTimestamp) UnmarshalJSON(data []byte) error {
	var tmp struct {
		LogicalTime int64  `json:"logicalTime"`
		WallTime    int64  `json:"wallTime"`
		NodeID      string `json:"nodeId"`
	}

	if err := json.Unmarshal(data, &tmp); err != nil {
		return err
	}

	t.LogicalTime = tmp.LogicalTime
	t.WallTime = tmp.WallTime
	t.NodeID = tmp.NodeID
	return nil
}

func max(a, b, c int64) int64 {
	result := a
	if b > result {
		result = b
	}
	if c > result {
		result = c
	}
	return result
}
